### PR TITLE
manifest: trusted-firmware-m: update to 2.1.0

### PR DESCRIPTION
**Pull request by tomi-font**
Update TF-M to version 2.1.0.

### CODE REVIEW

**Comment by tomi-font** [comment]
`psa-arch-tests` cannot currently be updated from 23.06 to 24.03 as some tests start to fail.
Upstream projects need to incorporate fixes for this update to be possible.

**Review by frkv — APPROVED** [review]
LGTM (pending passing tests and updates to correct SHA)

File: modules/trusted-firmware-m/Kconfig.tfm [hunk 1]
```diff
@@ -258,6 +237,28 @@ config TFM_CONNECTION_BASED_SERVICE_API
 
 if TFM_BL2
 
+config TFM_MCUBOOT_SIGNATURE_TYPE
+	string "The signature type used to sign the secure and non-secure firmware images."
+	default "EC-P256"
+	help
+	  Available types: RSA-2048, RSA-3072, EC-P256, EC-P384.
```
Comment (frkv) [inline modules/trusted-firmware-m/Kconfig.tfm:244]: Should ed25519 be on this list as well?

File: modules/trusted-firmware-m/CMakeLists.txt [hunk 2]
```diff
@@ -389,6 +401,8 @@ if (CONFIG_BUILD_WITH_TFM)
     if(PAD)
       set(pad_args --pad --pad-header)
     endif()
+    string(REPLACE "\${CONFIG_TFM_MCUBOOT_SIGNATURE_TYPE}" "${CONFIG_TFM_MCUBOOT_SIGNATURE_TYPE}"
+                    CONFIG_TFM_KEY_FILE_${SUFFIX} ${CONFIG_TFM_KEY_FILE_${SUFFIX}})
```
Comment (ceolin) [inline modules/trusted-firmware-m/CMakeLists.txt:405]: Are `CONFIG_TFM_KEY_FILE_S` and `CONFIG_TFM_KEY_FILE_NS` valid symbols ? Isn't TFM using `TFM_KEY_S` and `TFM_KEY_NS` ?

File: modules/trusted-firmware-m/CMakeLists.txt [hunk 3]
```diff
@@ -35,6 +35,18 @@ if (CONFIG_BUILD_WITH_TFM)
...
+
+    list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_SIGNATURE_TYPE=${CONFIG_TFM_MCUBOOT_SIGNATURE_TYPE})
+    if ("${CONFIG_TFM_MCUBOOT_SIGNATURE_TYPE}" MATCHES "^RSA")
+      list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_USE_PSA_CRYPTO=OFF)
+      list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_HW_KEY=ON)
+      list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_BUILTIN_KEY=OFF)
+    elseif ("${CONFIG_TFM_MCUBOOT_SIGNATURE_TYPE}" MATCHES "^EC-P")
+      list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_USE_PSA_CRYPTO=ON)
+      list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_HW_KEY=OFF)
+      list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_BUILTIN_KEY=ON)
```
Comment (ceolin) [inline modules/trusted-firmware-m/CMakeLists.txt:47]: Why these options based on the type of key ? the only dependency that is clear to me is `MCUBOOT_BUILTIN_KEY` requires `MCUBOOT_USE_PSA_CRYPTO` but I don't understand why, for example, you set HW_KEY for RSA and BUILTIN_KEY for EC-P.

Maybe just a comment explaining it would make things easier.

File: modules/trusted-firmware-m/Kconfig.tfm [hunk 4]
```diff
@@ -258,6 +237,28 @@ config TFM_CONNECTION_BASED_SERVICE_API
...
+	  that should be used by the BL2 bootloader when signing secure
+	  firmware images.
+
+config TFM_KEY_FILE_NS
+	string "Path to private key used to sign non-secure firmware images."
+	default "${ZEPHYR_TRUSTED_FIRMWARE_M_MODULE_DIR}/bl2/ext/mcuboot/root-${CONFIG_TFM_MCUBOOT_SIGNATURE_TYPE}_1.pem"
+	help
+	  The path and filename for the .pem file containing the private key
+	  that should be used by the BL2 bootloader when signing non-secure
+	  firmware images.
```
Comment (ceolin) [inline modules/trusted-firmware-m/Kconfig.tfm:260]: Are these symbols still used in TF-M or they were replaced ?

File: modules/trusted-firmware-m/Kconfig.tfm (hunk 1 above)
Comment (tomi-font) [reply modules/trusted-firmware-m/Kconfig.tfm:244]: No: https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/105/files#diff-c8402d394cb78c69fcf9bc110193858d9c6beb04277528dc74e30bcfe7a72e5bR50

File: modules/trusted-firmware-m/CMakeLists.txt (hunk 3 above)
Comment (tomi-font) [reply modules/trusted-firmware-m/CMakeLists.txt:47]: I added a comment to explain why.
RSA without `MCUBOOT_HW_KEY` and EC-P without `MCUBOOT_BUILTIN_KEY` failed to boot (`Image in the primary slot is not valid!` and `Provisioning failed`, respectively).

**Review by mswarowsky — APPROVED** [review]

**Review by Vge0rge — DISMISSED** [review]

**Review by mswarowsky — APPROVED** [review]
lgtm

**Review by frkv — APPROVED** [review]
LGTM

**Review by rlubos — DISMISSED** [review]

**Review by ceolin — APPROVED** [review]

File: samples/tfm_integration/psa_crypto/src/main.c [hunk 5]
```diff
@@ -44,5 +44,7 @@ int main(void)
 	/* Dump any queued log messages, and wait for system events. */
 	al_dump_log();
 
+	LOG_INF("Done.");
+
```
Comment (ithinuel) [inline samples/tfm_integration/psa_crypto/src/main.c:48]: Is this change required?

File: samples/tfm_integration/psa_crypto/src/psa_crypto.c [hunk 6]
```diff
@@ -510,7 +510,7 @@ void crp_generate_csr(void)
 	LOG_INF("Certificate Signing Request in JSON:\n");
 	al_dump_log();
 
-	printf("%s\n", json_encoded_buf);
+	printf("%s\n\n", json_encoded_buf);
```
Comment (ithinuel) [inline samples/tfm_integration/psa_crypto/src/psa_crypto.c:513]: Ditto

File: samples/tfm_integration/psa_crypto/src/main.c (hunk 5 above)
Comment (tomi-font) [reply samples/tfm_integration/psa_crypto/src/main.c:48]: To this and your other comment: Not strictly required as in the sample would be broken without them, but those very much help someone looking at the logs to understand that the sample is done instead of having hung. So to me yes, those are important enough to have.

**Review by ithinuel — APPROVED** [review]
